product_multi_barcode_constraint_per_company/models/product_barcode.py

```diff
@@ -0,0 +1,22 @@
+# Copyright 2023 Akretion (https://www.akretion.com).
+# @author Pierrick Brun <pierrick.brun@akretion.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class ProductBarcode(models.Model):
+    _inherit = "product.barcode"
+
+    def _get_duplicates(self, barcodes_to_check):
+        self.ensure_one()
+        barcodes_to_check = super()._get_duplicates(barcodes_to_check)
+        res = self.browse()
+        for barcode in barcodes_to_check:
+            if (
+                not barcode.company_id
+                or not self.company_id
+                or barcode.company_id == self.company_id
+            ):
+                res |= barcode
```

Comment:
@PierrickBrun for performances reasons:

```suggestion
            barcode_ids = set()
            for (...):
                barcode_ids.add(barcode.id)
            return self.browse(barcode_ids)
```

Reply:
Change done, this is the recordset modification that is bad performance-wise ?

Reply:
Yes, union is very slow.

Especially due to models name comparisons I think https://github.com/odoo/odoo/blob/16.0/odoo/models.py#L5790